augur/filter.py

```diff
@@ -141,16 +146,41 @@ def run(args):
     #If VCF, open and get sequence names
     if is_vcf:
         seq_keep, all_seq = read_vcf(args.sequences)
-
-    #if Fasta, read in file to get sequence names and sequences
     else:
-        try:
-            seqs = SeqIO.index(args.sequences, 'fasta')
-        except ValueError as error:
-            print("ERROR: Problem reading in {}:".format(args.sequences))
-            print(error)
-            return 1
-        seq_keep = list(seqs.keys())
+        # If FASTA, try to load the sequence composition details and strain
+        # names to be filtered.
+        index_is_autogenerated = False
+        sequence_index_path = args.sequence_index
+
+        # Generate the sequence index on the fly, for backwards compatibility
+        # with older workflows that don't generate the index ahead of time.
+        if sequence_index_path is None:
+            # Create a temporary index using a random filename to avoid
+            # collisions between multiple filter commands.
+            index_is_autogenerated = True
+            with NamedTemporaryFile(delete=False) as sequence_index_file:
+                sequence_index_path = sequence_index_file.name
+
+            print(
+                f"WARNING: A sequence index was not provided, so we are generating one.",
+                "Generate your own index ahead of time with `augur index` and pass it with `augur filter --sequence-index`.",
+                file=sys.stderr
+            )
+            index_sequences(args.sequences, sequence_index_path)
+
+        sequence_index = pd.read_csv(
+            sequence_index_path,
+            sep="\t"
+        )
+
+        # Remove temporary index file, if it exists.
+        if index_is_autogenerated:
+            os.unlink(sequence_index_path)
+
+        # Calculate summary statistics needed for filtering.
+        sequence_index["ACGT"] = sequence_index.loc[:, ["A", "C", "G", "T"]].sum(axis=1)
+        sequence_index["valid_nucleotides"] = sequence_index.loc[:, ["ACGT", "N", "other_IUPAC"]].sum(axis=1)
```

Comment:
Valid nucleotides should include "?" according to the original implementation of "non-nucleotide".